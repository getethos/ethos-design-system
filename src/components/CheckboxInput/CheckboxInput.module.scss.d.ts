// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'CheckboxInput': string;
  'Facade': string;
  'FacadeBorder': string;
  'checkboxWrapper': string;
  'checkboxWrapper2': string;
  'infoIcon': string;
  'label': string;
  'labelWithTooltip': string;
  'popper': string;
  'root': string;
  'root2': string;
  'tooltip': string;
  'tooltipBody': string;
}
export const cssExports: CssExports;
export default cssExports;
