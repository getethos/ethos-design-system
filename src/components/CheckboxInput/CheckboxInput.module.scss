.root {
  display: flex;
  align-items: flex-start;
}

.root2 {
  display: flex;
  align-items: stretch;
  min-height: var(--Space-56);
}

.label {
  flex: 1;
  display: flex;
  align-items: center;
  border: 1px solid rgba(0, 0, 0, 0.24);
  border-radius: 2px;
  cursor: pointer;
  transition: border-color background-color color 250ms ease;
  padding: 0px 16px;
}
.label input {
  cursor: inherit;
}
.CheckboxWrapper [data-disabled='true'] * {
  cursor: default;
}
.label > *:nth-child(2) {
  padding: 17px 0px 13px 0px;
}
.CheckboxWrapper [data-variant='textonly'][data-checked='true'] .label {
  background-color: var(--BrandForest);
  color: var(--White);
}

.CheckboxWrapper [data-checked='true'] label,
.CheckboxWrapper [data-checked='true'] .tooltip,
.CheckboxWrapper label:focus-within,
.CheckboxWrapper label:focus-within ~ .tooltip {
  border-color: var(--BrandForest);
}

.CheckboxWrapper [data-error='true'] label,
.CheckboxWrapper [data-error='true'] .tooltip,
.CheckboxWrapper [data-error='true'] label ~ .tooltip {
  border-color: var(--BrandSunburn);
}

.labelWithTooltip {
  border-radius: 2px 0px 0px 2px;
  border-right-width: 0px;
}

.checkboxWrapper {
  display: inline-block;
  position: relative;
  top: 1px;
  margin-right: 19px;
}

.checkboxWrapper2 {
  display: inline-block;
  position: relative;
  padding-right: var(--Space-16);
}
.CheckboxWrapper [data-variant='textonly'] .checkboxWrapper2 {
  clip: rect(0 0 0 0);
  clip-path: inset(50%);
  height: 1px;
  width: 1px;
  margin: 0;
  padding: 0;
  overflow: hidden;
}

.Facade {
  display: flex;
  align-items: center;
  justify-content: center;
  outline: 0;
  transition: border-color 250ms ease;
  width: 18px;
  height: 18px;
  overflow: hidden;
}

.FacadeBorder {
  border-color: var(--GrayStrokeAndDisabled--translucent);
  border-style: solid;
  border-width: 1px;
}

.FacadeWhiteBackground {
  background-color: white;
}

.FacadeDefaultBackground {
  background-color: transparent;
}

.CheckboxInput {
  position: absolute;
  opacity: 0;
}

/* State */
.CheckboxInput:not(:checked) ~ .Facade {
  &:hover {
    background-color: var(--GrayLightHover--translucent);
  }
}

.CheckboxInput:checked ~ .Facade {
  background-color: var(--BrandForest);
  border-width: 0px;
  opacity: 1;
  fill: var(--white);
}

/**
 * Hackâ€”this ensures that our Errors.module.scss does not get overriden by this rule
 * (normal focus causes dark gray, while the error still results in red border)
 */
.CheckboxInput:focus ~ .Facade:not(:global(.FacadeError)) {
  border-color: var(--GrayPrimary--opaque);
}

.tooltip,
.tooltip * {
  display: flex;
  transition: border-color 250ms ease;
}

.tooltip {
  background: #f5f5f5;
  border-top-width: 1px;
  border-right-width: 1px;
  border-bottom-width: 1px;
  border-left-width: 0px;
  border-color: #c2c2c2;
  border-style: solid;
  border-radius: 0px 2px 2px 0px;
}

.tooltipBody {
  display: flex;
  align-items: center;
  padding: 8px;
  border-left: 1px solid #c2c2c2;
}
[data-variant='textonly'][data-checked='true'] .tooltipBody {
  border-left-color: var(--BrandForest);
}

.popper {
  border-radius: 2px;
  background-color: var(--GrayPrimary--opaque);
  color: var(--White);
  left: var(--Space-4) !important;

  ::before {
    border-color: transparent var(--GrayPrimary--opaque) transparent transparent !important;
  }
}

.infoIcon {
  display: block;
  width: var(--Space-16);
  height: var(--Space-16);
}
