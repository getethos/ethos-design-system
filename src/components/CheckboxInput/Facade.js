import React from 'react'
import PropTypes from 'prop-types'

export const Facade = ({ className }) => {
  return (
    <svg
      className={className}
      width="18"
      height="18"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.75 12.127L3.623 9l-1.065 1.057L6.75 14.25l9-9-1.057-1.058-7.943 7.935z"
        fill="#fff"
      />
    </svg>
  )
}
Facade.propTypes = {
  className: PropTypes.string.isRequired,
}
