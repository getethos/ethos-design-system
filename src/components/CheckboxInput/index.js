export { CheckboxInput } from './CheckboxInput.js'
