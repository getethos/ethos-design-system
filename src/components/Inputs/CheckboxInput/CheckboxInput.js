import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Body } from '../../Type/Body.js'
import useErrorMessage from '../../../hooks/useErrorMessage.js'
import useInputValidation from '../../../hooks/useInputValidation.js'
import styles from './CheckboxInput.module.scss'
import errorStyles from '../Errors.module.scss'

const Facade = () => {
  return (
    <svg
      className={styles.Facade}
      width="18"
      height="18"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.75 12.127L3.623 9l-1.065 1.057L6.75 14.25l9-9-1.057-1.058-7.943 7.935z"
        fill="#fff"
      />
    </svg>
  )
}

export const CheckboxInput = ({
  formChangeHandler,
  validator,
  children,
  optional,
  name,
  initialValue,
  currentValue,
  currentError,
  formTouched,
  ...rest
}) => {

  const initialChecked = currentValue || initialValue || false 
  const [touched, setTouched] = useState(initialChecked)
  const [isChecked, setIsChecked] = useState(initialChecked)
  const [getError, setError, getFormattedError, validate] = useErrorMessage(validator)
  const [doValidation] = useInputValidation({validate, setError, formChangeHandler})

  const onChange = (ev) => {
    // It feels like a checkbox isn't something you blur off of, so I'm electing to 
    // call a checkbox touched when you check it
    if (!touched) setTouched(true)
    const val = ev.target.type === 'checkbox' ? ev.target.checked : ev.target.value;
    doValidation(val, touched)
    setIsChecked(val)
  }

  const onKeyPress = (ev) => {
    if (ev.key === 'Space' || ev.keyCode === 32) {
      const val = ev.target.type === 'checkbox' ? ev.target.checked : ev.target.value;
      doValidation(!val, touched)
      setIsChecked(!val)
    }
  }

  const getClasses = () => {
    return !!getError(currentError, touched) ? `${styles.CheckboxInput} ${errorStyles.Error}` : `${styles.CheckboxInput}`
  }

  const id = name
  const otherProps = { ...rest, id, name }

  return (
    <>
      <label htmlFor={id} className={styles.root}>
        <div className={styles.checkboxWrapper}>
          <input
            className={styles.CheckboxInput}
            type="checkbox"
            onChange={onChange}
            onKeyPress={onKeyPress}
            checked={isChecked}
            {...otherProps}
          />
          <Facade />
        </div>
        <Body.Regular400>{children}</Body.Regular400>
      </label>
      {getError(currentError, formTouched)}
    </>
  )
}

CheckboxInput.propTypes = {
  optional: PropTypes.bool,
  name: PropTypes.string.isRequired, // must be unique
  'data-tid': PropTypes.string.isRequired,
  initialValue: PropTypes.bool,
  children: PropTypes.node.isRequired,
  disabled: PropTypes.bool,
  allCaps: PropTypes.bool,
  validator: PropTypes.func,
  formChangeHandler: PropTypes.func,
}
