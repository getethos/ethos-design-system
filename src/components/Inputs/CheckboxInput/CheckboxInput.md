```jsx
// formChangeHandler gets wired up automatically if using <Form /> component
import validateExists from '../../../validators/validateExists'
const formChangeHandlerStub = () => {}
;<CheckboxInput
  name="le-check-unchecked"
  data-tid="le-tid-unchecked"
  formChangeHandler={formChangeHandlerStub}
  validator={validateExists}
>
  I agree to the{' '}
  <a href="/" target="_blank">
    Agreement
  </a>{' '}
  , and the{' '}
  <a href="/" target="_blank">
    Other Agreement
  </a>
</CheckboxInput>
```

Set to checked and disabled immediately:

```jsx
<CheckboxInput
  initialValue={true}
  name="le-check2"
  disabled={true}
  data-tid="le-tid2"
>
  I answered for you...mwa ha ha. I agree to the{' '}
  <a href="/" target="_blank">
    Agreement
  </a>
</CheckboxInput>
```

This one sets an `intialValue` which results in the field being considered as
already `touched`. This means you do not have to `blur` for field hint error
messages to appear. _Note, you can think of initialValue as "checked" if that
helpsâ€”true results in a checked checkbox, false results in unchecked._

```jsx
import validateExists from '../../../validators/validateExists'
const formChangeHandlerStub = () => {}

;<CheckboxInput
  initialValue={true}
  name="le-check3"
  data-tid="le-tid-checked"
  formChangeHandler={formChangeHandlerStub}
  validator={validateExists}
>
  I agree to the{' '}
  <a href="/" target="_blank">
    Agreement
  </a>
</CheckboxInput>
```
