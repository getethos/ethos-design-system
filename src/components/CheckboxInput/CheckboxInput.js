import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { COLORS } from '../Colors.js'
import { Body } from '../Body.js'
import useErrorMessage from '../../hooks/useErrorMessage.js'
import useInputValidation from '../../hooks/useInputValidation.js'
import { codes } from '../../helpers/constants.js'
import { Facade } from './Facade.js'
import styles from './CheckboxInput.module.scss'
import errorStyles from '../Errors.module.scss'

export const CheckboxInput = ({
  formChangeHandler,
  validator,
  children,
  disabled,
  name,
  initialValue,
  currentValue,
  currentError,
  setFieldTouched,
  // TODO: pushed in from Form.js but not used here
  formTouched, // eslint-disable-line no-unused-vars
  checked,
  ...rest
}) => {
  const initialChecked = currentValue || initialValue || false
  const [touched, setTouched] = useState(initialValue ? true : false)
  const [isChecked, setIsChecked] = useState(initialChecked)

  const [getError, setError, , validate] = useErrorMessage(validator)
  const [doValidation] = useInputValidation({
    validate,
    setError,
    formChangeHandler,
  })

  // Initial Value aka prefilledâ€”are considered "touched", but must prevalidate
  // which will in turn update the internal form state as to their validity
  useEffect(() => {
    if (!!formChangeHandler && initialValue) {
      doValidation(initialValue, true)
    }
  }, [])

  const onChange = (ev) => {
    // It feels like a checkbox isn't something you blur off of, so I'm electing to
    // call a checkbox touched when you check it
    if (!touched) setTouched(true)
    const val =
      ev.target.type === 'checkbox' ? ev.target.checked : ev.target.value
    doValidation(val, touched)
    setIsChecked(val)
    if (setFieldTouched) {
      setFieldTouched(true)
    }
  }

  const onKeyPress = (ev) => {
    if (ev.keyCode === codes.SPACE) {
      const val =
        ev.target.type === 'checkbox' ? ev.target.checked : ev.target.value
      doValidation(!val, touched)
      setIsChecked(!val)
      if (setFieldTouched) {
        setFieldTouched(true)
      }
    }
  }

  const getFacade = () => {
    const klasses = getFacadeClasses()
    if (rest.facadeRenderer) {
      return rest.facadeRenderer(klasses)
    } else {
      return <Facade classes={klasses} />
    }
  }
  const getFacadeClasses = () => {
    return getError(currentError, touched)
      ? `${styles.Facade} FacadeError ${errorStyles.Error}`
      : `${styles.Facade} ${styles.FacadeBorder}`
  }

  const getClasses = () => {
    return getError(currentError, touched)
      ? `${styles.CheckboxInput} ${errorStyles.Error}`
      : `${styles.CheckboxInput}`
  }

  /**
   * If consumer's used the brute force `checked` prop, we prioritize that
   */
  const resolvedIsChecked = typeof checked !== 'undefined' ? checked : isChecked

  return (
    <>
      <label htmlFor={name} className={styles.root}>
        <div className={styles.checkboxWrapper}>
          <input
            className={getClasses()}
            type="checkbox"
            onChange={onChange}
            data-tid={rest['data-tid']}
            onKeyPress={onKeyPress}
            disabled={disabled}
            checked={resolvedIsChecked}
            id={name}
            name={name}
          />
          {getFacade()}
        </div>
        <Body.Regular400 color={COLORS.GRAY_PRIMARY}>
          {children}
        </Body.Regular400>
      </label>
      {getError(currentError, touched)}
    </>
  )
}

CheckboxInput.propTypes = {
  /** low level prop used only by the form engine */
  formTouched: PropTypes.bool,
  /** Name of the field */
  name: PropTypes.string.isRequired,
  /** Required data-tid used as a unique id for targeting test selectors */
  'data-tid': PropTypes.string.isRequired,
  /** Optionally sets a default value for the checkbox */
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  /** Allows for brute force setting of whether checked or not. Usually,
   * you should let the component manage this itself, but this may be
   * useful if you need to set the checked state programatically */
  checked: PropTypes.bool,
  /** currentValue is a low level prop used only by the form engine */
  currentValue: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  /** setFieldTouched is a low level prop used only by the form engine */
  setFieldTouched: PropTypes.func,
  /** currentError is a low level prop used only by the form engine */
  currentError: PropTypes.string,
  /** formChangeHandler is a low level prop used only by the form engine */
  formChangeHandler: PropTypes.func,
  /** The Accordion's children. Likely AccordionSection's */
  children: PropTypes.node.isRequired,
  /** `disabled` - whether to disable the the checkbox */
  disabled: PropTypes.bool,
  /** whether to display label in all caps */
  allCaps: PropTypes.bool,
  /** text transform capitalize label */
  capitalize: PropTypes.bool,
  /** field validation callback */
  validator: PropTypes.func,
  /** only useful if you need to completely override the checkbox facade e.g
   * NoraCheckboxInput does this */
  facadeRenderer: PropTypes.func,
}
