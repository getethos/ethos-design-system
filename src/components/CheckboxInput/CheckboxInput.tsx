import * as React from 'react'
import { useEffect, useState } from 'react'
import { codes } from '../../helpers/constants.js'
import useErrorMessage from '../../hooks/useErrorMessage'
import useInputValidation from '../../hooks/useInputValidation'
import { Body } from '../Body.js'
import { COLORS } from '../Colors.js'
import errorStyles from '../Errors.module.scss'
import { PublicFormProps } from '../Form/Form'
import styles from './CheckboxInput.module.scss'
import { Facade } from './Facade'

type CheckboxInputProps = PublicFormProps & {
  initialValue?: string | boolean
  checked?: boolean
  currentValue?: string | boolean
  currentError?: string
  disabled?: boolean
  facadeRenderer?: (...args: any[]) => any
}
export const CheckboxInput: React.SFC<CheckboxInputProps> = ({
  formChangeHandler,
  validator,
  children,
  disabled,
  name,
  initialValue,
  currentValue,
  currentError,
  setFieldTouched,
  // TODO: pushed in from Form.js but not used here
  formTouched, // eslint-disable-line no-unused-vars
  checked,
  ...rest
}) => {
  const initialChecked = currentValue || initialValue || false
  const [touched, setTouched] = useState(initialValue ? true : false)
  const [isChecked, setIsChecked] = useState(initialChecked)
  const [getError, setError, , validate] = useErrorMessage(validator)
  const [doValidation] = useInputValidation({
    validate,
    setError,
    formChangeHandler,
  })
  // Initial Value aka prefilledâ€”are considered "touched", but must prevalidate
  // which will in turn update the internal form state as to their validity
  useEffect(() => {
    if (!!formChangeHandler && initialValue) {
      doValidation(initialValue, true)
    }
  }, [])
  const onChange = (ev) => {
    // It feels like a checkbox isn't something you blur off of, so I'm electing to
    // call a checkbox touched when you check it
    if (!touched) setTouched(true)
    const val =
      ev.target.type === 'checkbox' ? ev.target.checked : ev.target.value
    doValidation(val, touched)
    setIsChecked(val)
    if (setFieldTouched) {
      setFieldTouched(true)
    }
  }
  const onKeyPress = (ev) => {
    if (ev.keyCode === codes.SPACE) {
      const val =
        ev.target.type === 'checkbox' ? ev.target.checked : ev.target.value
      doValidation(!val, touched)
      setIsChecked(!val)
      if (setFieldTouched) {
        setFieldTouched(true)
      }
    }
  }
  const getFacade = () => {
    const klasses = getFacadeClasses()
    if (rest.facadeRenderer) {
      return rest.facadeRenderer(klasses)
    } else {
      return <Facade classes={klasses} />
    }
  }
  const getFacadeClasses = () => {
    return getError(currentError, touched)
      ? `${styles.Facade} FacadeError ${errorStyles.Error}`
      : `${styles.Facade} ${styles.FacadeBorder}`
  }
  const getClasses = () => {
    return getError(currentError, touched)
      ? `${styles.CheckboxInput} ${errorStyles.Error}`
      : `${styles.CheckboxInput}`
  }
  /**
   * If consumer's used the brute force `checked` prop, we prioritize that
   */
  const resolvedIsChecked = typeof checked !== 'undefined' ? checked : isChecked
  return (
    <>
      <label htmlFor={name} className={styles.root}>
        <div className={styles.checkboxWrapper}>
          <input
            className={getClasses()}
            type="checkbox"
            onChange={onChange}
            data-tid={rest['data-tid']}
            onKeyPress={onKeyPress}
            disabled={disabled}
            checked={!!resolvedIsChecked}
            id={name}
            name={name}
          />
          {getFacade()}
        </div>
        <Body.Regular400 color={COLORS.GRAY_PRIMARY}>
          {children}
        </Body.Regular400>
      </label>
      {getError(currentError, touched)}
    </>
  )
}
