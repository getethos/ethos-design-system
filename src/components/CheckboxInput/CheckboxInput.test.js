import React from 'react'
import { CheckboxInput } from './CheckboxInput'
import renderer from 'react-test-renderer'

const formChangeHandlerStub = jest.fn()

describe('CheckboxInput', () => {
  it('default rendering', () => {
    const tree = renderer
      .create(
        <CheckboxInput
          name="le-check"
          data-tid="le-tid"
          formChangeHandler={formChangeHandlerStub}
          validator={(x) => (x === true ? '' : 'Nope!')}
        >
          I agree to the{' '}
          <a href="/" target="_blank">
            Agreement
          </a>
        </CheckboxInput>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('checked takes precedence over initial and current values', () => {
    const rendered = renderer.create(
      <CheckboxInput
        // Note `currentValue` is generally only passed in by the form engine, but
        // we do it here because we need to ensure `checked` is taking precedence.
        currentValue={false}
        initialValue={false}
        checked={true}
        name="le-check"
        data-tid="le-tid"
      >
        le children
      </CheckboxInput>
    )
    const tree = rendered.toTree()
    expect(tree.props.checked).toBe(true)
    expect(rendered.toJSON()).toMatchSnapshot()
  })
})
