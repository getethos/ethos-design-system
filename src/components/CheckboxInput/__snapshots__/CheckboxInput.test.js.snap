// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxInput checked takes precedence over initial and current values 1`] = `
Array [
  <label
    className="root"
    htmlFor="le-check"
  >
    <div
      className="checkboxWrapper"
    >
      <input
        checked={true}
        className="CheckboxInput"
        data-tid="le-tid"
        id="le-check"
        name="le-check"
        onChange={[Function]}
        onKeyPress={[Function]}
        type="checkbox"
      />
      <svg
        className="Facade FacadeDefaultBackground FacadeBorder"
        fill="none"
        height="18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.75 12.127L3.623 9l-1.065 1.057L6.75 14.25l9-9-1.057-1.058-7.943 7.935z"
          fill="#fff"
        />
      </svg>
    </div>
    <div
      className="Body Theinhardt Regular400 GrayPrimary"
    >
      le children
    </div>
  </label>,
  "",
]
`;

exports[`CheckboxInput default rendering 1`] = `
Array [
  <label
    className="root"
    htmlFor="le-check"
  >
    <div
      className="checkboxWrapper"
    >
      <input
        checked={false}
        className="CheckboxInput"
        data-tid="le-tid"
        id="le-check"
        name="le-check"
        onChange={[Function]}
        onKeyPress={[Function]}
        type="checkbox"
      />
      <svg
        className="Facade FacadeDefaultBackground FacadeBorder"
        fill="none"
        height="18"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.75 12.127L3.623 9l-1.065 1.057L6.75 14.25l9-9-1.057-1.058-7.943 7.935z"
          fill="#fff"
        />
      </svg>
    </div>
    <div
      className="Body Theinhardt Regular400 GrayPrimary"
    >
      I agree to the
       
      <a
        href="/"
        target="_blank"
      >
        Agreement
      </a>
    </div>
  </label>,
  "",
]
`;
